File: realtime-macos/Sources/PAPAluLive/CharacterDefinition.swift
import AppKit

enum CharacterID: String, Codable, Equatable, Hashable {
    case catMeme
    case huhCat
    case happyCat
    case screamingCat
    case papalu
    case custom
}

struct BundledCharacterDefinition: Equatable {
    let definition: CharacterDefinition
    let resourceDirectoryName: String
}

struct CharacterFrameStep: Equatable {
    let assetName: String
    let duration: Double
}

struct IdleMotionConfiguration: Equatable {
    let horizontalOffset: Double
    let rotationDegrees: Double
    let durationRange: ClosedRange<Double>
    let holdRange: ClosedRange<Double>

    static let gentle = IdleMotionConfiguration(
        horizontalOffset: 4,
        rotationDegrees: 1,
        durationRange: 0.95...1.15,
        holdRange: 0.08...0.25
    )
}

struct CharacterDefinition: Equatable {
    let id: CharacterID
    let name: String
    let idleAssetName: String
    let talkingAssetNames: [String]
    let talkingFramesPerSecond: Double
    let blinkSteps: [CharacterFrameStep]
    let settleSteps: [CharacterFrameStep]
    let blinkDelayRange: ClosedRange<Double>?
    let idleMotion: IdleMotionConfiguration
    let thoughtCloudEnabled: Bool
    let defaultSize: NSSize

    static let catMeme = twoFrameBuiltIn(
        id: .catMeme,
        name: "猫 Meme"
    )

    static let huhCat = twoFrameBuiltIn(
        id: .huhCat,
        name: "Huh 猫"
    )

    static let happyCat = twoFrameBuiltIn(
        id: .happyCat,
        name: "Happy 猫"
    )

    static let screamingCat = twoFrameBuiltIn(
        id: .screamingCat,
        name: "抱头尖叫猫"
    )

    static let papalu = CharacterDefinition(
        id: .papalu,
        name: "PAPAlu",
        idleAssetName: "0",
        talkingAssetNames: ["2", "1", "3", "4", "6", "3"],
        talkingFramesPerSecond: 8,
        blinkSteps: [
            CharacterFrameStep(assetName: "5", duration: 0.11),
            CharacterFrameStep(assetName: "7", duration: 0.10),
            CharacterFrameStep(assetName: "0", duration: 0.12),
        ],
        settleSteps: ["3", "1", "7", "0"].map {
            CharacterFrameStep(assetName: $0, duration: 0.08)
        },
        blinkDelayRange: 3...5,
        idleMotion: .gentle,
        thoughtCloudEnabled: true,
        defaultSize: NSSize(width: 288, height: 312)
    )

    static func custom(name: String) -> CharacterDefinition {
        CharacterDefinition(
            id: .custom,
            name: name,
            idleAssetName: "idle",
            talkingAssetNames: [
                "talking", "idle", "talking", "idle", "talking", "talking",
            ],
            talkingFramesPerSecond: 8,
            blinkSteps: [],
            settleSteps: [],
            blinkDelayRange: nil,
            idleMotion: .gentle,
            thoughtCloudEnabled: true,
            defaultSize: NSSize(width: 288, height: 312)
        )
    }

    static let bundledCharacters = [
        BundledCharacterDefinition(
            definition: .catMeme,
            resourceDirectoryName: "CatMeme"
        ),
        BundledCharacterDefinition(
            definition: .huhCat,
            resourceDirectoryName: "HuhCat"
        ),
        BundledCharacterDefinition(
            definition: .happyCat,
            resourceDirectoryName: "HappyCat"
        ),
        BundledCharacterDefinition(
            definition: .screamingCat,
            resourceDirectoryName: "ScreamingCat"
        ),
        BundledCharacterDefinition(
            definition: .papalu,
            resourceDirectoryName: "PAPAlu"
        ),
    ]

    private static func twoFrameBuiltIn(
        id: CharacterID,
        name: String
    ) -> CharacterDefinition {
        CharacterDefinition(
            id: id,
            name: name,
            idleAssetName: "idle",
            talkingAssetNames: [
                "talking", "idle", "talking", "idle", "talking", "talking",
            ],
            talkingFramesPerSecond: 8,
            blinkSteps: [],
            settleSteps: [
                CharacterFrameStep(assetName: "talking", duration: 0.08),
                CharacterFrameStep(assetName: "idle", duration: 0.08),
            ],
            blinkDelayRange: nil,
            idleMotion: .gentle,
            thoughtCloudEnabled: true,
            defaultSize: NSSize(width: 288, height: 312)
        )
    }
}

File: realtime-windows/PAPAluLive.Core/CharacterDefinition.cs
namespace PAPAluLive.Core;

public enum CharacterId
{
    CatMeme,
    HuhCat,
    HappyCat,
    ScreamingCat,
    Papalu,
    Custom,
}

public enum CharacterDisplayState
{
    Idle,
    Talking,
}

public sealed record FrameStep(string AssetName, double Duration);

public sealed record CharacterSize(double Width, double Height);

public sealed record IdleMotionConfiguration(
    double HorizontalOffset,
    double RotationDegrees,
    double MinimumDuration,
    double MaximumDuration,
    double MinimumHold,
    double MaximumHold)
{
    public static IdleMotionConfiguration Gentle { get; } = new(
        HorizontalOffset: 4,
        RotationDegrees: 1,
        MinimumDuration: 0.95,
        MaximumDuration: 1.15,
        MinimumHold: 0.08,
        MaximumHold: 0.25);
}

public sealed record CharacterDefinition(
    CharacterId Id,
    string Name,
    string ResourceDirectoryName,
    string IdleAssetName,
    IReadOnlyList<string> TalkingAssetNames,
    double TalkingFramesPerSecond,
    IReadOnlyList<FrameStep> BlinkSteps,
    IReadOnlyList<FrameStep> SettleSteps,
    double? MinimumBlinkDelay,
    double? MaximumBlinkDelay,
    IdleMotionConfiguration IdleMotion,
    bool ThoughtCloudEnabled,
    CharacterSize DefaultSize)
{
    public static CharacterDefinition CatMeme { get; } = TwoFrameBuiltIn(
        CharacterId.CatMeme,
        "猫 Meme",
        "CatMeme");

    public static CharacterDefinition HuhCat { get; } = TwoFrameBuiltIn(
        CharacterId.HuhCat,
        "Huh 猫",
        "HuhCat");

    public static CharacterDefinition HappyCat { get; } = TwoFrameBuiltIn(
        CharacterId.HappyCat,
        "Happy 猫",
        "HappyCat");

    public static CharacterDefinition ScreamingCat { get; } = TwoFrameBuiltIn(
        CharacterId.ScreamingCat,
        "抱头尖叫猫",
        "ScreamingCat");

    public static CharacterDefinition Papalu { get; } = new(
        Id: CharacterId.Papalu,
        Name: "PAPAlu",
        ResourceDirectoryName: "PAPAlu",
        IdleAssetName: "0",
        TalkingAssetNames: ["2", "1", "3", "4", "6", "3"],
        TalkingFramesPerSecond: 8,
        BlinkSteps:
        [
            new FrameStep("5", 0.11),
            new FrameStep("7", 0.10),
            new FrameStep("0", 0.12),
        ],
        SettleSteps:
        [
            new FrameStep("3", 0.08),
            new FrameStep("1", 0.08),
            new FrameStep("7", 0.08),
            new FrameStep("0", 0.08),
        ],
        MinimumBlinkDelay: 3,
        MaximumBlinkDelay: 5,
        IdleMotion: IdleMotionConfiguration.Gentle,
        ThoughtCloudEnabled: true,
        DefaultSize: new CharacterSize(288, 312));

    public static IReadOnlyList<CharacterDefinition> BundledCharacters { get; } =
    [
        CatMeme,
        HuhCat,
        HappyCat,
        ScreamingCat,
        Papalu,
    ];

    public static CharacterDefinition Custom(string name = "自定义角色") => new(
        Id: CharacterId.Custom,
        Name: name,
        ResourceDirectoryName: "CustomCharacter",
        IdleAssetName: "idle",
        TalkingAssetNames: ["talking"],
        TalkingFramesPerSecond: 1,
        BlinkSteps: [],
        SettleSteps: [],
        MinimumBlinkDelay: null,
        MaximumBlinkDelay: null,
        IdleMotion: IdleMotionConfiguration.Gentle,
        ThoughtCloudEnabled: false,
        DefaultSize: new CharacterSize(288, 312));

    public IReadOnlySet<string> RequiredAssetNames => new HashSet<string>(
        new[] { IdleAssetName }
            .Concat(TalkingAssetNames)
            .Concat(BlinkSteps.Select(step => step.AssetName))
            .Concat(SettleSteps.Select(step => step.AssetName)));

    private static CharacterDefinition TwoFrameBuiltIn(
        CharacterId id,
        string name,
        string resourceDirectoryName) => new(
            Id: id,
            Name: name,
            ResourceDirectoryName: resourceDirectoryName,
            IdleAssetName: "idle",
            TalkingAssetNames:
            [
                "talking",
                "idle",
                "talking",
                "idle",
                "talking",
                "talking",
            ],
            TalkingFramesPerSecond: 8,
            BlinkSteps: [],
            SettleSteps:
            [
                new FrameStep("talking", 0.08),
                new FrameStep("idle", 0.08),
            ],
            MinimumBlinkDelay: null,
            MaximumBlinkDelay: null,
            IdleMotion: IdleMotionConfiguration.Gentle,
            ThoughtCloudEnabled: true,
            DefaultSize: new CharacterSize(288, 312));
}
